liquidator/fasta_scorer.cpp

```diff
@@ -0,0 +1,292 @@
+        output_mutex(other.output_mutex)
+    {}
+
+    ~Scorer()
+    {
+        std::lock_guard<std::mutex> lock(output_mutex);
+
+        for (const auto& info: infos)
+        {
+            output << info.motif_name << '\t';
+            output.write(fasta.data() + info.fasta_sequence_name_begin, info.fasta_sequence_name_length);
+            output << '\t'
+                   << static_cast<unsigned>(info.sequence_start) << '\t'
+                   << static_cast<unsigned>(info.sequence_stop) << '\t'
+                   << info.strand << '\t';
+
+            output.precision(6);
+            output << info.score << '\t';
+            output.precision(3);
+
+            output << info.pvalue << '\t'
+                   << '\t'; // omit q-value for now
+
+            // todo: should really do toupper
+            // sequence_begin +1 for new line between name end and sequence, -1 for start being 1 based (so cancels out)
+            const size_t sequence_begin = info.fasta_sequence_name_begin + info.fasta_sequence_name_length + info.sequence_start;
+
+            // sequence_length +1 because start is 1 based
+            const size_t sequence_length = info.sequence_stop - info.sequence_start + 1;
+            if (info.strand == '+')
+            {
+                output.write(fasta.data() + sequence_begin,
+                             sequence_length);
+            }
+            else
+            {
+                for (size_t i=sequence_begin + sequence_length - 1; ; --i)
+                {
+                    output << complement(fasta[i]);
+                    if (i==sequence_begin)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            output << std::endl;
+        }
+    }
+
+    Scorer& operator=(const Scorer& other) = delete;
+
+    void score(size_t region_begin, const size_t region_end)
+    {
+        while(true)
+        {
+            // todo: is > a valid char in a sequence name? maybe we should search for "\n>" char* instead of single char
+            size_t name_begin = fasta.find_first_of('>', region_begin);
+            if (name_begin >= region_end)
+            {
+                break;
+            }
+            name_begin++; // name starts character after '>'
+            size_t name_end = fasta.find_first_of('\n', name_begin+1);
+
+            size_t sequence_begin = name_end + 1;
+            // this assumes the file is terminated by a newline, otherwise last sequence won't have an end
+            size_t sequence_end = fasta.find_first_of('\n', sequence_begin + 1);
+
+            for (const auto& matrix : matrices)
+            {
+                for (size_t begin = sequence_begin, end = sequence_begin + matrix.matrix().size();
+                     end <= sequence_end;
+                     ++begin, ++end)
+                {
+                    // todo: experiment with switching loop order: move matrix loop inside sequence loop
```

Comment:
When two matrices of the same size (this is the case when single motif, since both forward and reverse matrices), reversing the loop order slightly decreased scoring performance by about 5%.